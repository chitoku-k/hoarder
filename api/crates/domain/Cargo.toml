[package]
name = "domain"
authors.workspace = true
edition.workspace = true
license.workspace = true

[lib]
doctest = false

[dependencies.chrono]
version = "0.4.39"
features = ["clock", "std"]
default-features = false

[dependencies.derive_more]
version = "2.0.1"
features = ["constructor", "deref", "display", "from"]

[dependencies.futures]
version = "0.3.31"
features = ["std"]
default-features = false

[dependencies.indenter]
version = "0.3.3"

[dependencies.ordermap]
version = "0.5.5"

[dependencies.percent-encoding]
version = "2.3.1"

[dependencies.regex]
version = "1.11.1"

[dependencies.serde]
version = "1.0.218"
features = ["derive"]

[dependencies.strum]
version = "0.27.0"
features = ["derive"]

[dependencies.thiserror]
version = "2.0.11"

[dependencies.tokio]
version = "1.43.0"
features = ["macros", "rt-multi-thread"]

[dependencies.tracing]
version = "0.1.41"

[dependencies.uuid]
version = "1.13.1"
features = ["serde", "v4"]

[dev-dependencies.anyhow]
version = "1.0.95"

[dev-dependencies.indoc]
version = "2.0.5"

[dev-dependencies.mockall]
version = "0.13.1"

[dev-dependencies.pretty_assertions]
version = "1.4.1"
features = ["unstable"]

[dev-dependencies.serial_test]
version = "3.2.0"
