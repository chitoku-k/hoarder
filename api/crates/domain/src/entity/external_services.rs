use std::{borrow::Cow, str::FromStr};

use derive_more::{Deref, Display, From};
use regex::Regex;
use serde::{Deserialize, Serialize};
use strum::{EnumString, IntoStaticStr};
use uuid::Uuid;

#[derive(Clone, Copy, Debug, Default, Deref, Deserialize, Display, Eq, From, Hash, Ord, PartialEq, PartialOrd, Serialize)]
pub struct ExternalServiceId(Uuid);

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ExternalService {
    pub id: ExternalServiceId,
    pub slug: String,
    pub kind: ExternalServiceKind,
    pub name: String,
    pub base_url: Option<String>,
    pub url_pattern: Option<String>,
}

impl ExternalService {
    pub fn metadata_by_url(&self, url: &str) -> Option<ExternalMetadata> {
        let (id, creator_id) = self.url_pattern
            .as_ref()
            .and_then(|url_pattern| Regex::new(url_pattern).ok())
            .and_then(|re| re.captures(url))
            .map(|captures| {
                let id = captures.name("id").map(|c| c.as_str());
                let creator_id = captures.name("creatorId").map(|c| c.as_str());
                (id, creator_id)
            })
            .unwrap_or_default();

        ExternalMetadata::from_metadata(&self.kind, url, id, creator_id)
    }
}

#[derive(Clone, Debug, strum::Display, EnumString, Eq, PartialEq)]
#[strum(serialize_all = "snake_case")]
pub enum ExternalServiceKind {
    Bluesky,
    Fantia,
    Mastodon,
    Misskey,
    Nijie,
    Pixiv,
    PixivFanbox,
    Pleroma,
    Seiga,
    Skeb,
    Threads,
    Website,
    X,
    Xfolio,
    #[strum(default)]
    Custom(String),
}

impl ExternalServiceKind {
    pub fn default_base_urls(&self) -> &'static [&'static str] {
        use ExternalServiceKind::*;
        match self {
            Bluesky => &["https://bsky.app"],
            Fantia => &["https://fantia.jp"],
            Nijie => &["https://nijie.info"],
            Pixiv => &["https://www.pixiv.net"],
            Seiga => &["https://seiga.nicovideo.jp"],
            Skeb => &["https://skeb.jp"],
            Threads => &["https://www.threads.net"],
            X => &["https://x.com", "https://twitter.com"],
            Xfolio => &["https://xfolio.jp"],
            _ => &[],
        }
    }

    pub fn default_url_pattern(&self, base_urls: &[&str]) -> Option<String> {
        use ExternalServiceKind::*;
        let (prefixes, pattern): (&[_], _) = match self {
            Bluesky => (&[], r"/profile/(?<creatorId>[^/?#]+)/post/(?<id>[^/?#]+)(?:[?#].*)?$"),
            Fantia => (&[], r"/posts/(?<id>\d+)(?:[?#].*)?$"),
            Mastodon => (&[], r"/@(?<creatorId>[^/?#]+)/(?<id>\d+)(?:[?#].*)?$"),
            Misskey => (&[], r"/notes/(?<id>[^/?#]+)(?:[?#].*)?$"),
            Nijie => (&[], r"/view\.php\?id=(?<id>\d+)(?:[?#].*)?$"),
            Pixiv => (&[], r"/(?:artworks/|member_illust\.php\?(?:|.+&)illust_id=)(?<id>\d+)(?:[?&#].*)?$"),
            PixivFanbox => (&[r"(?<creatorId>[^.]+)\.fanbox\.cc", r"www\.fanbox\.cc/@(?:[^.]+)"], r"/posts/(?<id>\d+)(?:[?#].*)?$"),
            Pleroma => (&[], r"/notice/(?<id>[^/?#]+)(?:[?#].*)?$"),
            Seiga => (&[], r"/seiga/(?<id>\d+)(?:[?#].*)?$"),
            Skeb => (&[], r"/@(?<creatorId>[^/]+)/works/(?<id>\d+)(?:[?#].*)?$"),
            Threads => (&[], r"/(?<creatorId>[^/]+)/post/(?<id>[^/$#]+)(?:[?#].*)?$"),
            X => (&[], r"/(?<creatorId>[^/]+)/status/(?<id>\d+)(?:[/?#].*)?$"),
            Xfolio => (&[], r"/portfolio/(?<creatorId>[^/]+)/works/(?<id>\d+)(?:[?#].*)?$"),
            Website | Custom(..) => return None,
        };

        let hostnames = prefixes
            .iter()
            .map(|&prefix| Cow::from(prefix))
            .chain(base_urls
                .iter()
                .map(|base_url| Cow::from(
                    regex::escape(base_url
                        .trim_start_matches("http://")
                        .trim_start_matches("https://")
                        .trim_end_matches("/")))))
            .reduce(|mut acc, s| {
                acc.to_mut().push('|');
                acc.to_mut().push_str(&s);
                acc
            })?;

        if prefixes.len() + base_urls.len() == 1 {
            Some(format!("^https?://{hostnames}{pattern}"))
        } else {
            Some(format!("^https?://(?:{hostnames}){pattern}"))
        }
    }
}

impl From<String> for ExternalServiceKind {
    fn from(value: String) -> Self {
        Self::try_from(value.as_str()).unwrap()
    }
}

#[derive(Clone, Debug, Eq, IntoStaticStr, PartialEq)]
#[strum(serialize_all = "snake_case")]
pub enum ExternalMetadata {
    Bluesky { id: String, creator_id: String },
    Fantia { id: u64 },
    Mastodon { id: u64, creator_id: String },
    Misskey { id: String },
    Nijie { id: u64 },
    Pixiv { id: u64 },
    PixivFanbox { id: u64, creator_id: String },
    Pleroma { id: String },
    Seiga { id: u64 },
    Skeb { id: u64, creator_id: String },
    Threads { id: String, creator_id: Option<String> },
    Website { url: String },
    X { id: u64, creator_id: Option<String> },
    Xfolio { id: u64, creator_id: String },
    Custom(String),
}

impl ExternalMetadata {
    pub fn from_metadata(kind: &ExternalServiceKind, url: &str, id: Option<&str>, creator_id: Option<&str>) -> Option<Self> {
        use ExternalServiceKind::*;
        let metadata = match kind {
            Bluesky => Self::Bluesky { id: id?.to_string(), creator_id: creator_id?.to_string() },
            Fantia => Self::Fantia { id: id?.parse().ok()? },
            Mastodon => Self::Mastodon { id: id?.parse().ok()?, creator_id: creator_id?.to_string() },
            Misskey => Self::Misskey { id: id?.to_string() },
            Nijie => Self::Nijie { id: id?.parse().ok()? },
            Pixiv => Self::Pixiv { id: id?.parse().ok()? },
            PixivFanbox => Self::PixivFanbox { id: id?.parse().ok()?, creator_id: creator_id?.to_string() },
            Pleroma => Self::Pleroma { id: id?.to_string() },
            Seiga => Self::Seiga { id: id?.parse().ok()? },
            Skeb => Self::Skeb { id: id?.parse().ok()?, creator_id: creator_id?.to_string() },
            Threads => Self::Threads { id: id?.to_string(), creator_id: creator_id.map(Into::into) },
            Website => Self::Website { url: url.to_string() },
            X => Self::X { id: id?.parse().ok()?, creator_id: creator_id.map(Into::into) },
            Xfolio => Self::Xfolio { id: id?.parse().ok()?, creator_id: creator_id?.to_string() },
            Custom(..) => return None,
        };

        Some(metadata)
    }

    pub fn kind(&self) -> Option<ExternalServiceKind> {
        if let ExternalMetadata::Custom(_) = self {
            None
        } else {
            ExternalServiceKind::from_str(self.into()).ok()
        }
    }

    pub fn url(&self, base_url: Option<&str>) -> Option<Cow<str>> {
        let base_url = base_url.map(|b| b.trim_end_matches("/"));

        use ExternalMetadata::*;
        let url = match self {
            Bluesky { id, creator_id } => format!("{}/profile/{creator_id}/post/{id}", base_url?).into(),
            Fantia { id } => format!("{}/posts/{id}", base_url?).into(),
            Mastodon { id, creator_id } => format!("{}/@{creator_id}/{id}", base_url?).into(),
            Misskey { id } => format!("{}/notes/{id}", base_url?).into(),
            Nijie { id } => format!("{}/view.php?id={id}", base_url?).into(),
            Pixiv { id } => format!("{}/artworks/{id}", base_url?).into(),
            PixivFanbox { id, creator_id } => format!("https://{creator_id}.fanbox.cc/posts/{id}").into(),
            Pleroma { id } => format!("{}/notice/{id}", base_url?).into(),
            Seiga { id } => format!("{}/seiga/im{id}", base_url?).into(),
            Skeb { id, creator_id } => format!("{}/@{creator_id}/works/{id}", base_url?).into(),
            Threads { id, creator_id } => format!("{}/@{}/post/{id}", base_url?, creator_id.as_deref().unwrap_or_default()).into(),
            Website { url } => url.into(),
            X { id, creator_id } => format!("{}/{}/status/{id}", base_url?, creator_id.as_deref().unwrap_or("i")).into(),
            Xfolio { id, creator_id } => format!("{}/portfolio/{creator_id}/works/{id}", base_url?).into(),
            Custom(..) => return None,
        };

        Some(url)
    }
}
