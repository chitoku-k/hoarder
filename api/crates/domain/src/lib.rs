#![allow(clippy::too_many_arguments)]

pub mod entity;
pub mod parser;
pub mod processor;
pub mod repository;
pub mod service;
