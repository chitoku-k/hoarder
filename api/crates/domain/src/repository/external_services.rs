use std::future::Future;

use crate::{
    entity::external_services::{ExternalService, ExternalServiceId},
    error::Result,
    repository::DeleteResult
};

pub trait ExternalServicesRepository: Send + Sync + 'static {
    /// Creates an external service.
    fn create(&self, slug: &str, kind: &str, name: &str, base_url: Option<&str>, url_pattern: Option<&str>) -> impl Future<Output = Result<ExternalService>> + Send;

    /// Fetches the external services by their IDs.
    fn fetch_by_ids<T>(&self, ids: T) -> impl Future<Output = Result<Vec<ExternalService>>> + Send
    where
        for<'a> T: IntoIterator<Item = ExternalServiceId> + Send + 'a;

    /// Fetches all external services.
    fn fetch_all(&self) -> impl Future<Output = Result<Vec<ExternalService>>> + Send;

    /// Updates the external service by ID.
    fn update_by_id(&self, id: ExternalServiceId, slug: Option<&str>, name: Option<&str>, base_url: Option<Option<&str>>, url_pattern: Option<Option<&str>>) -> impl Future<Output = Result<ExternalService>> + Send;

    /// Deletes the external service by ID.
    fn delete_by_id(&self, id: ExternalServiceId) -> impl Future<Output = Result<DeleteResult>> + Send;
}
