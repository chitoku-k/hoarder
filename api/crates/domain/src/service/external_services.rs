use std::future::Future;

use crate::error::Result;

use derive_more::Constructor;

use crate::{
    entity::external_services::{ExternalService, ExternalServiceId},
    repository::{external_services, DeleteResult},
};

#[cfg_attr(feature = "test-mock", mockall::automock)]
pub trait ExternalServicesServiceInterface: Send + Sync + 'static {
    /// Creates an external service.
    fn create_external_service<'a>(&self, slug: &str, kind: &str, name: &str, base_url: Option<&'a str>) -> impl Future<Output = Result<ExternalService>> + Send;

    /// Gets external services.
    fn get_external_services(&self) -> impl Future<Output = Result<Vec<ExternalService>>> + Send;

    /// Gets the external services by their IDs.
    fn get_external_services_by_ids<T>(&self, ids: T) -> impl Future<Output = Result<Vec<ExternalService>>> + Send
    where
        T: IntoIterator<Item = ExternalServiceId> + Send + Sync + 'static;

    /// Updates the external service by ID.
    fn update_external_service_by_id<'a>(&self, id: ExternalServiceId, slug: Option<&'a str>, name: Option<&'a str>, base_url: Option<Option<&'a str>>) -> impl Future<Output = Result<ExternalService>> + Send;

    /// Deletes the external service by ID.
    fn delete_external_service_by_id(&self, id: ExternalServiceId) -> impl Future<Output = Result<DeleteResult>> + Send;
}

#[derive(Clone, Constructor)]
pub struct ExternalServicesService<ExternalServicesRepository> {
    external_services_repository: ExternalServicesRepository,
}

impl<ExternalServicesRepository> ExternalServicesServiceInterface for ExternalServicesService<ExternalServicesRepository>
where
    ExternalServicesRepository: external_services::ExternalServicesRepository,
{
    async fn create_external_service<'a>(&self, slug: &str, kind: &str, name: &str, base_url: Option<&'a str>) -> Result<ExternalService> {
        match self.external_services_repository.create(slug, kind, name, base_url).await {
            Ok(service) => Ok(service),
            Err(e) => {
                log::error!("failed to create an external service\nError: {e:?}");
                Err(e)
            },
        }
    }

    async fn get_external_services(&self) -> Result<Vec<ExternalService>> {
        match self.external_services_repository.fetch_all().await {
            Ok(services) => Ok(services),
            Err(e) => {
                log::error!("failed to get external services\nError: {e:?}");
                Err(e)
            },
        }
    }

    async fn get_external_services_by_ids<T>(&self, ids: T) -> Result<Vec<ExternalService>>
    where
        T: IntoIterator<Item = ExternalServiceId> + Send + Sync + 'static,
    {
        match self.external_services_repository.fetch_by_ids(ids).await {
            Ok(services) => Ok(services),
            Err(e) => {
                log::error!("failed to get external services\nError: {e:?}");
                Err(e)
            },
        }
    }

    async fn update_external_service_by_id<'a>(&self, id: ExternalServiceId, slug: Option<&'a str>, name: Option<&'a str>, base_url: Option<Option<&'a str>>) -> Result<ExternalService> {
        match self.external_services_repository.update_by_id(id, slug, name, base_url).await {
            Ok(service) => Ok(service),
            Err(e) => {
                log::error!("failed to update the external service\nError: {e:?}");
                Err(e)
            },
        }
    }

    async fn delete_external_service_by_id(&self, id: ExternalServiceId) -> Result<DeleteResult> {
        match self.external_services_repository.delete_by_id(id).await {
            Ok(result) => Ok(result),
            Err(e) => {
                log::error!("failed to delete the external service\nError: {e:?}");
                Err(e)
            },
        }
    }
}
