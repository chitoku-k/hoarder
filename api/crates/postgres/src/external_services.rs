use derive_more::{Constructor, From, Into};
use domain::{
    entity::external_services::{ExternalService, ExternalServiceId},
    error::{Error, ErrorKind, Result},
    repository::{external_services::ExternalServicesRepository, DeleteResult},
};
use futures::TryStreamExt;
use sea_query::{Expr, Iden, LockType, Order, PostgresQueryBuilder, Query};
use sea_query_binder::SqlxBinder;
use sqlx::{FromRow, PgPool};

use crate::sea_query_uuid_value;

#[derive(Clone, Constructor)]
pub struct PostgresExternalServicesRepository {
    pool: PgPool,
}

#[derive(Clone, Debug, From, Into)]
pub(crate) struct PostgresExternalServiceId(ExternalServiceId);

#[derive(Debug, FromRow)]
pub(crate) struct PostgresExternalServiceRow {
    id: PostgresExternalServiceId,
    slug: String,
    name: String,
}

#[derive(Iden)]
pub(crate) enum PostgresExternalService {
    #[iden = "external_services"]
    Table,
    Id,
    Slug,
    Name,
}

sea_query_uuid_value!(PostgresExternalServiceId, ExternalServiceId);

impl From<PostgresExternalServiceRow> for ExternalService {
    fn from(row: PostgresExternalServiceRow) -> Self {
        Self {
            id: row.id.into(),
            slug: row.slug,
            name: row.name,
        }
    }
}

impl ExternalServicesRepository for PostgresExternalServicesRepository {
    async fn create(&self, slug: &str, name: &str) -> Result<ExternalService> {
        let (sql, values) = Query::insert()
            .into_table(PostgresExternalService::Table)
            .columns([
                PostgresExternalService::Slug,
                PostgresExternalService::Name,
            ])
            .values([
                slug.into(),
                name.into(),
            ])
            .map_err(Error::other)?
            .returning(
                Query::returning()
                    .columns([
                        PostgresExternalService::Id,
                        PostgresExternalService::Slug,
                        PostgresExternalService::Name,
                    ])
            )
            .build_sqlx(PostgresQueryBuilder);

        match sqlx::query_as_with::<_, PostgresExternalServiceRow, _>(&sql, values).fetch_one(&self.pool).await {
            Ok(external_service) => Ok(external_service.into()),
            Err(sqlx::Error::Database(e)) if e.is_unique_violation() => Err(ErrorKind::ExternalServiceSlugDuplicate { slug: slug.to_string() })?,
            Err(e) => Err(Error::other(e)),
        }
    }

    async fn fetch_by_ids<T>(&self, ids: T) -> Result<Vec<ExternalService>>
    where
        T: IntoIterator<Item = ExternalServiceId> + Send + Sync,
    {
        let (sql, values) = Query::select()
            .columns([
                PostgresExternalService::Id,
                PostgresExternalService::Slug,
                PostgresExternalService::Name,
            ])
            .from(PostgresExternalService::Table)
            .and_where(Expr::col(PostgresExternalService::Id).is_in(ids.into_iter().map(PostgresExternalServiceId::from)))
            .order_by(PostgresExternalService::Slug, Order::Asc)
            .build_sqlx(PostgresQueryBuilder);

        let external_services = sqlx::query_as_with::<_, PostgresExternalServiceRow, _>(&sql, values)
            .fetch(&self.pool)
            .map_ok(Into::into)
            .try_collect()
            .await
            .map_err(Error::other)?;

        Ok(external_services)
    }

    async fn fetch_all(&self) -> Result<Vec<ExternalService>> {
        let (sql, values) = Query::select()
            .columns([
                PostgresExternalService::Id,
                PostgresExternalService::Slug,
                PostgresExternalService::Name,
            ])
            .from(PostgresExternalService::Table)
            .order_by(PostgresExternalService::Slug, Order::Asc)
            .build_sqlx(PostgresQueryBuilder);

        let external_services = sqlx::query_as_with::<_, PostgresExternalServiceRow, _>(&sql, values)
            .fetch(&self.pool)
            .map_ok(Into::into)
            .try_collect()
            .await
            .map_err(Error::other)?;

        Ok(external_services)
    }

    async fn update_by_id<'a>(&self, id: ExternalServiceId, name: Option<&'a str>) -> Result<ExternalService> {
        let mut tx = self.pool.begin().await.map_err(Error::other)?;

        let (sql, values) = Query::select()
            .columns([
                PostgresExternalService::Id,
                PostgresExternalService::Slug,
                PostgresExternalService::Name,
            ])
            .from(PostgresExternalService::Table)
            .and_where(Expr::col(PostgresExternalService::Id).eq(PostgresExternalServiceId::from(id)))
            .lock(LockType::Update)
            .build_sqlx(PostgresQueryBuilder);

        let external_service = match sqlx::query_as_with::<_, PostgresExternalServiceRow, _>(&sql, values).fetch_one(&mut *tx).await {
            Ok(row) => ExternalService::from(row),
            Err(sqlx::Error::RowNotFound) => return Err(ErrorKind::ExternalServiceNotFound { id })?,
            Err(e) => return Err(Error::other(e)),
        };

        let name = name.unwrap_or(&external_service.name);

        let (sql, values) = Query::update()
            .table(PostgresExternalService::Table)
            .value(PostgresExternalService::Name, name)
            .and_where(Expr::col(PostgresExternalService::Id).eq(PostgresExternalServiceId::from(id)))
            .returning(
                Query::returning()
                    .columns([
                        PostgresExternalService::Id,
                        PostgresExternalService::Slug,
                        PostgresExternalService::Name,
                    ])
            )
            .build_sqlx(PostgresQueryBuilder);

        let external_service = match sqlx::query_as_with::<_, PostgresExternalServiceRow, _>(&sql, values).fetch_one(&mut *tx).await {
            Ok(row) => ExternalService::from(row),
            Err(e) => return Err(Error::other(e)),
        };

        tx.commit().await.map_err(Error::other)?;
        Ok(external_service)
    }

    async fn delete_by_id(&self, id: ExternalServiceId) -> Result<DeleteResult> {
        let (sql, values) = Query::delete()
            .from_table(PostgresExternalService::Table)
            .and_where(Expr::col(PostgresExternalService::Id).eq(PostgresExternalServiceId::from(id)))
            .build_sqlx(PostgresQueryBuilder);

        let affected = sqlx::query_with(&sql, values)
            .execute(&self.pool)
            .await
            .map_err(Error::other)?
            .rows_affected();

        match affected {
            0 => Ok(DeleteResult::NotFound),
            count => Ok(DeleteResult::Deleted(count)),
        }
    }
}
