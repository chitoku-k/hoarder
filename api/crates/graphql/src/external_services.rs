use async_graphql::SimpleObject;
use domain::entity::external_services;
use uuid::Uuid;

#[derive(SimpleObject)]
pub(crate) struct ExternalService {
    id: Uuid,
    slug: String,
    kind: String,
    name: String,
    base_url: Option<String>,
    url_pattern: Option<String>,
}

impl From<external_services::ExternalService> for ExternalService {
    fn from(external_service: external_services::ExternalService) -> Self {
        Self {
            id: *external_service.id,
            slug: external_service.slug,
            kind: external_service.kind,
            name: external_service.name,
            base_url: external_service.base_url,
            url_pattern: external_service.url_pattern,
        }
    }
}
