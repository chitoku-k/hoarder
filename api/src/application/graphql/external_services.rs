use async_graphql::SimpleObject;
use uuid::Uuid;

use crate::domain::entity::external_services;

#[derive(SimpleObject)]
pub struct ExternalService {
    id: Uuid,
    slug: String,
    name: String,
}

impl From<external_services::ExternalService> for ExternalService {
    fn from(external_service: external_services::ExternalService) -> Self {
        Self {
            id: *external_service.id,
            slug: external_service.slug,
            name: external_service.name,
        }
    }
}
