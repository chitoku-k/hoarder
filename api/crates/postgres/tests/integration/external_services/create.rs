use domain::{
    entity::external_services::ExternalServiceKind,
    error::ErrorKind,
    repository::external_services::ExternalServicesRepository,
};
use insta::assert_toml_snapshot;
use postgres::external_services::PostgresExternalServicesRepository;
use pretty_assertions::{assert_eq, assert_matches};
use sqlx::Row;
use test_context::test_context;
use tracing::Instrument;

use super::DatabaseContext;

#[test_context(DatabaseContext)]
#[tokio::test]
async fn succeeds(ctx: &DatabaseContext) {
    let repository = PostgresExternalServicesRepository::new(ctx.pool.clone());
    let actual = repository.create(
        "foobar",
        ExternalServiceKind::Custom("foobar".to_string()),
        "FooBar",
        Some("https://foobar.example.com"),
        Some(r"^https?://foobar\.example\.com/(?<id>\d+)(?:[/?#].*)?$"),
    ).instrument(ctx.span.clone()).await.unwrap();

    assert_eq!(actual.slug, "foobar".to_string());
    assert_eq!(actual.kind, ExternalServiceKind::Custom("foobar".to_string()));
    assert_eq!(actual.name, "FooBar".to_string());
    assert_eq!(actual.base_url, Some("https://foobar.example.com".to_string()));
    assert_eq!(actual.url_pattern, Some(r"^https?://foobar\.example\.com/(?<id>\d+)(?:[/?#].*)?$".to_string()));

    let actual = sqlx::query(r#"SELECT "id", "slug", "kind", "name", "base_url", "url_pattern" FROM "external_services" WHERE "id" = $1"#)
        .bind(*actual.id)
        .fetch_one(&ctx.pool)
        .await
        .unwrap();

    assert_eq!(actual.get::<&str, &str>("slug"), "foobar");
    assert_eq!(actual.get::<&str, &str>("kind"), "foobar");
    assert_eq!(actual.get::<&str, &str>("name"), "FooBar");
    assert_eq!(actual.get::<Option<&str>, &str>("base_url"), Some("https://foobar.example.com"));
    assert_eq!(actual.get::<Option<&str>, &str>("url_pattern"), Some(r"^https?://foobar\.example\.com/(?<id>\d+)(?:[/?#].*)?$"));

    assert_toml_snapshot!(ctx.queries());
}

#[test_context(DatabaseContext)]
#[tokio::test]
async fn fails(ctx: &DatabaseContext) {
    let repository = PostgresExternalServicesRepository::new(ctx.pool.clone());
    let actual = repository.create(
        "x",
        ExternalServiceKind::X,
        "X",
        Some("https://x.com"),
        Some(r"^https?://foobar\.example\.com/(?<id>\d+)(?:[/?#].*)?$"),
    ).instrument(ctx.span.clone()).await.unwrap_err();

    assert_matches!(actual.kind(), ErrorKind::ExternalServiceSlugDuplicate { slug } if slug == "x");

    assert_toml_snapshot!(ctx.queries());
}
