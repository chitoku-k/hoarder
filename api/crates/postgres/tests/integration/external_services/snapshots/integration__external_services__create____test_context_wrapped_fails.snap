---
source: crates/postgres/tests/integration/external_services/create.rs
expression: ctx.queries()
---
[[queries]]
sql = '''
INSERT INTO
  "external_services" (
    "slug",
    "kind",
    "name",
    "base_url",
    "url_pattern"
  )
VALUES
  ($1, $2, $3, $4, $5)
RETURNING
  "id",
  "slug",
  "kind",
  "name",
  "base_url",
  "url_pattern"'''
rows_affected = 0
rows_returned = 0
